File: lang/src/main/java/org/xvm/lsp/model/CompilationResult.java
package org.xvm.lsp.model;

import org.jspecify.annotations.NonNull;

import java.util.List;

/**
 * Immutable result of compiling an XTC source file.
 * This is what the adapter layer produces from the compiler.
 */
public record CompilationResult(
        @NonNull String uri,
        @NonNull List<Diagnostic> diagnostics,
        @NonNull List<SymbolInfo> symbols,
        boolean success
) {
    public CompilationResult {
        diagnostics = List.copyOf(diagnostics);
        symbols = List.copyOf(symbols);
    }

    public static CompilationResult success(
            final @NonNull String uri,
            final @NonNull List<SymbolInfo> symbols) {
        return new CompilationResult(uri, List.of(), symbols, true);
    }

    public static CompilationResult withDiagnostics(
            final @NonNull String uri,
            final @NonNull List<Diagnostic> diagnostics,
            final @NonNull List<SymbolInfo> symbols) {
        final boolean hasErrors = diagnostics.stream()
                .anyMatch(d -> d.severity() == Diagnostic.Severity.ERROR);
        return new CompilationResult(uri, diagnostics, symbols, !hasErrors);
    }

    public static CompilationResult failure(
            final @NonNull String uri,
            final @NonNull List<Diagnostic> diagnostics) {
        return new CompilationResult(uri, diagnostics, List.of(), false);
    }
}

File: lang/lsp-server/src/main/kotlin/org/xvm/lsp/model/CompilationResult.kt
package org.xvm.lsp.model

/**
 * Immutable result of compiling an XTC source file.
 * This is what the adapter layer produces from the compiler.
 */
data class CompilationResult(
    val uri: String,
    val diagnostics: List<Diagnostic>,
    val symbols: List<SymbolInfo>,
    val success: Boolean,
    val documentUris: Set<String> = setOf(uri),
) {
    companion object {
        fun success(
            uri: String,
            symbols: List<SymbolInfo>,
        ): CompilationResult = CompilationResult(uri, emptyList(), symbols.toList(), true)

        fun withDiagnostics(
            uri: String,
            diagnostics: List<Diagnostic>,
            symbols: List<SymbolInfo>,
            documentUris: Set<String> = setOf(uri),
        ): CompilationResult {
            val hasErrors = diagnostics.any { it.severity == Diagnostic.Severity.ERROR }
            return CompilationResult(uri, diagnostics.toList(), symbols.toList(), !hasErrors, documentUris.toSet())
        }

        fun failure(
            uri: String,
            diagnostics: List<Diagnostic>,
        ): CompilationResult = CompilationResult(uri, diagnostics.toList(), emptyList(), false)
    }
}
